src/core/ddsi/src/ddsi_typelookup.c

```diff
@@ -0,0 +1,356 @@
+  serdata->timestamp = ddsrt_time_wallclock ();
+
+  GVTRACE (" wr-tl-req "PGUIDFMT" typeid "PTYPEIDFMT"\n", PGUID (wr->e.guid), PTYPEID(*type_id));
+  struct ddsi_tkmap_instance *tk = ddsi_tkmap_lookup_instance_ref (gv->m_tkmap, serdata);
+  write_sample_gc (lookup_thread_state (), NULL, wr, serdata, tk);
+  tlm->state = TL_META_REQUESTED;
+  ddsi_tkmap_instance_unref (gv->m_tkmap, tk);
+  ddsrt_mutex_unlock (&gv->tl_admin_lock);
+  return true;
+}
+
+static void write_typelookup_reply (struct participant *pp, seqno_t seqno, type_identifier_type_object_pair_seq_t *types)
+{
+  struct ddsi_domaingv * const gv = pp->e.gv;
+  type_lookup_reply_t reply;
+
+  struct writer *wr = get_builtin_writer (pp, NN_ENTITYID_TL_SVC_BUILTIN_REPLY_WRITER);
+  if (wr == NULL)
+  {
+    GVTRACE ("write_typelookup_reply("PGUIDFMT") - builtin TypeLookup service reply writer not found\n", PGUID (pp->e.guid));
+    return;
+  }
+
+  memcpy (&reply.writer_guid, &wr->e.guid, sizeof (wr->e.guid));
+  reply.sequence_number = seqno;
+  reply.types.n = types->n;
+  reply.types.types = types->types;
+  struct ddsi_serdata *serdata = ddsi_serdata_from_sample (gv->tl_svc_reply_type, SDK_DATA, &reply);
+  serdata->timestamp = ddsrt_time_wallclock ();
+
+  GVTRACE (" wr-tl-reply wr "PGUIDFMT"\n", PGUID (wr->e.guid));
+  struct ddsi_tkmap_instance *tk = ddsi_tkmap_lookup_instance_ref (gv->m_tkmap, serdata);
+  write_sample_gc (lookup_thread_state (), NULL, wr, serdata, tk);
+  ddsi_tkmap_instance_unref (gv->m_tkmap, tk);
+}
+
+void ddsi_tl_handle_request (struct ddsi_domaingv *gv, const ddsi_guid_prefix_t *guid_prefix, struct ddsi_serdata *sample_common)
+{
+  assert (!(sample_common->statusinfo & (NN_STATUSINFO_DISPOSE | NN_STATUSINFO_UNREGISTER)));
+  const struct ddsi_serdata_pserop *sample = (const struct ddsi_serdata_pserop *) sample_common;
+  const type_lookup_request_t *req = sample->sample;
+  GVTRACE (" handle-tl-req wr "PGUIDFMT " seqnr %"PRIi64" ntypeids %"PRIu32, PGUID (req->writer_guid), req->sequence_number, req->type_ids.n);
+
+  ddsrt_mutex_lock (&gv->tl_admin_lock);
+  type_identifier_type_object_pair_seq_t types = { 0, NULL };
+  for (uint32_t n = 0; n < req->type_ids.n; n++)
+  {
+    type_identifier_t *tid = &req->type_ids.type_ids[n];
+    GVTRACE (" type "PTYPEIDFMT, PTYPEID (*tid));
+    struct tl_meta *tlm = ddsi_tl_meta_lookup_locked (gv, tid);
+    if (tlm != NULL && tlm->state == TL_META_RESOLVED)
+    {
+      size_t sz;
+      types.n++;
+      types.types = ddsrt_realloc (types.types, types.n * sizeof (*types.types));
+      memcpy (&types.types[types.n - 1].type_identifier, tlm->type_id, sizeof (types.types[types.n - 1].type_identifier));
+      tlm->sertype->ops->serialize (tlm->sertype, &sz, &types.types[types.n - 1].type_object.value);
+      assert (sz <= UINT32_MAX);
+      types.types[types.n - 1].type_object.length = (uint32_t) sz;
+      GVTRACE (" found");
+    }
+  }
+  ddsrt_mutex_unlock (&gv->tl_admin_lock);
+
+  if (types.n > 0)
```

Comment:
I'd think a reply should be written also if there is no substance in the response: that way the requestor doesn't have to timeout before it can continue.

Reply:
I agree, removed the count > 0 condition here, so it'll always reply to an incoming type lookup request (unless there is no typelookup reply writer, which should of course not happen where the is a request reader).